cli/src/cli_util.rs

```diff
@@ -692,20 +692,81 @@ impl CommandHelper {
+            .ok_or_else(|| CommandError::InternalError("Could not load commit ID".to_string()))?;
+        let commit = repo.store().get_commit(commit_id)?;
+        let tree_id = commit.tree_id();
+
+        let mut tx = repo.start_transaction(&self.settings);
+        let mut_repo = tx.mut_repo();
+        let new_commit = mut_repo
+            .new_commit(&self.settings, vec![commit_id.clone()], tree_id.clone())
+            .write()?;
+        mut_repo.set_wc_commit(workspace.workspace_id().clone(), new_commit.id().clone())?;
+        let ret = Ok(tx.commit("recovery commit"));
+
+        let mut locked_workspace = workspace.start_working_copy_mutation()?;
+        locked_workspace
+            .locked_wc()
+            .reset(&repo.store().get_root_tree(tree_id)?)
+            //.check_out(&new_commit)
+            .map_err(|err| {
+                CommandError::InternalError(format!(
+                    "Could not check out commit in workspace: {err}"
+                ))
+            })?;
+        locked_workspace.finish(operation.id().clone())?;
+
+        writeln!(
+            ui.stderr(),
+            "Created and checked out recovery commit {}",
+            new_commit.id().hex()
+        )?;
+
+        ret
+    }
+
     /// Loads workspace that will diverge from the last working-copy operation.
     pub fn for_stale_working_copy(
         &self,
         ui: &mut Ui,
-    ) -> Result<WorkspaceCommandHelper, CommandError> {
-        let workspace = self.load_workspace()?;
+    ) -> Result<(WorkspaceCommandHelper, bool), CommandError> {
+        let mut workspace = self.load_workspace()?;
         let op_store = workspace.repo_loader().op_store();
-        let op_id = workspace.working_copy().operation_id();
-        let op_data = op_store
-            .read_operation(op_id)
-            .map_err(|e| CommandError::InternalError(format!("Failed to read operation: {e}")))?;
-        let operation = Operation::new(op_store.clone(), op_id.clone(), op_data);
-        let repo = workspace.repo_loader().load_at(&operation)?;
-        self.for_loaded_repo(ui, workspace, repo)
+        let (repo, recovered) = {
+            let op_id = workspace.working_copy().operation_id();
+            match op_store.read_operation(op_id) {
+                Ok(op_data) => (
+                    workspace.repo_loader().load_at(&Operation::new(
+                        op_store.clone(),
+                        op_id.clone(),
+                        op_data,
+                    ))?,
+                    false,
+                ),
+                Err(e) => {
+                    writeln!(
+                        ui.stderr(),
+                        "Failed to read working copy's current operation; attempting recovery. \
+                         Error message from read attempt: {e}"
+                    )?;
+                    (
+                        self.create_and_check_out_recovery_commit(ui, &mut workspace)?,
```

Comment:
Perhaps, this fallback (and `create_and_check_out_recovery_commit()`) can be moved to the caller if this recovery method is an explicit option (something like `jj workspace update-stale --new-wc-commit`.)

It might be a bit inconvenient, but I think that's okay so long as the situation is rare.

Reply:
I was thinking about that too - I think that this whole function (for_stale_working_copy) should be moved to the caller, actually. cli_util.rs seems to be getting too large. If people have consensus, I can take a look at that.

Reply:
Ah yes, the original `for_stale_working_copy()` can also be moved to the caller.

Anyway, I prefer an explicit option to do `create_and_check_out_recovery_commit()` because the end result would look different, and it's valid to use that option to recover from the other stale situations.

Reply:
OK, I've refactored `for_stale_working_copy()` so it's now in the caller. As for the explicit option, since it's now moved to the caller, I don't think it's needed (because the caller always wants it).